File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/cobol-canonical-rollstock/src/ConsumeRollService.cs
using System.Threading;
using System.Threading.Tasks;

namespace Demo.RollStock;

/// <summary>
/// Roll-stock consumption service derived from a SIGNED COBOL spec for
/// CONSUME-ROLL. The COBOL claim taxonomy maps to this C# shape as:
///   INV-* (invariants)         → behavioural guards in this method
///   SC-* (section contracts)   → method signature + result enum
///   IO-* (I/O side effects)    → IRollRepository + IEventNotifier
///   EC-* (edge cases)          → branch-by-branch tests
///   Q-*  (open questions)      → MUST be resolved before signature
/// Method body is stubbed; implementation completes per the cited
/// invariants in the signed spec.
/// </summary>
public sealed class ConsumeRollService
{
    // INV-5 magic constant — surfaced as a named field so the
    // implementation can swap to a per-grade lookup if Q-2 resolves.
    public const decimal MinRemainLf = 12.0m;

    private readonly IRollRepository _rolls;
    private readonly IEventNotifier _events;

    public ConsumeRollService(IRollRepository rolls, IEventNotifier events)
    {
        _rolls = rolls;
        _events = events;
    }

    public Task<ConsumeRollResult> ConsumeAsync(
        string rollId,
        decimal usedLf,
        string operatorId,
        CancellationToken ct = default)
    {
        // TODO: implement per the signed COBOL spec
        //   INV-1 / IO-1: VSAM READ on ROLL-MASTER; not-found → RESULT_CD=1
        //   INV-2: locked rolls return RESULT_CD=3 without WRITE
        //   INV-3: USED_LF > ON_HAND_LF returns RESULT_CD=2
        //   INV-4: NEW_LF = ON_HAND_LF - USED_LF (no clamping)
        //   INV-5: NEW_LF < MIN_REMAIN sets ROLL_STATUS = 9 (Depleted)
        //   INV-6 / IO-2: success path performs VSAM REWRITE then
        //                 emits INV_CHG via IEventNotifier
        throw new System.NotImplementedException(
            "Engineer-implementation required against the signed COBOL spec.");
    }
}

public enum ConsumeRollResult
{
    Ok = 0,
    NotFound = 1,
    Insufficient = 2,
    Locked = 3,
}

File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-rollstock/src/main/java/com/example/rollstock/ConsumeRollService.java
package com.example.rollstock;

/**
 * Roll-stock consumption service derived from the signed spec for
 * CONSUME_ROLL. Method body is stubbed; implementation completes per
 * the cited spec invariants.
 *
 * <p>This is the Spring-flavoured mirror of the .NET 8 archetype's
 * <code>ConsumeRollService.cs</code>. Claim ids (INV-*, SE-*, EC-*)
 * map 1:1 to the signed spec — every TODO names which invariant it
 * unblocks.</p>
 */
public final class ConsumeRollService {

    /** Spec INV-5 magic constant — surfaced as a named field. */
    public static final java.math.BigDecimal MIN_REMAIN_LF = new java.math.BigDecimal("12.0");

    private final RollRepository rolls;
    private final EventNotifier events;

    public ConsumeRollService(RollRepository rolls, EventNotifier events) {
        this.rolls = rolls;
        this.events = events;
    }

    /**
     * TODO: implement per signed spec
     *   INV-1: RESULT_CD=1 (NotFound) when the read yields not-found
     *   INV-2: locked rolls return Locked without modification
     *   INV-3: usedLf > onHandLf returns Insufficient
     *   INV-4: newLf = onHandLf - usedLf (no clamping)
     *   INV-5: newLf < MIN_REMAIN sets rollStatus = 9 (Depleted)
     *   INV-6: success path writes + emits an event
     */
    public ConsumeRollResult consume(
            String rollId,
            java.math.BigDecimal usedLf,
            String operatorId) {
        throw new UnsupportedOperationException(
            "See signed spec INV-1..6 — engineer implementation required.");
    }
}
